csrc/val_graph.h

```diff
@@ -0,0 +1,216 @@
+
+  // Return if there's a group entry in the graph for this val
+  bool hasGroup(Val* val) const;
+
+  // Convert expr to its exprGroup, assert that it exists.
+  const ExprGroup& toGroup(Expr* expr) const;
+
+  // Convert Val to its ValGroup, assert that it exists.
+  const ValGroup& toGroup(Val* val) const;
+
+  // Return output/input Val groups of provided expr
+  // Note that the same ValGroup can show up multiple times, so the
+  // output type cannot be VectorOfUniqueEntries
+  std::vector<ValGroup> outputGroups(const ExprGroup& expr) const;
+  std::vector<ValGroup> inputGroups(const ExprGroup& expr) const;
+
+  // Recursively traverses uses of the IdGroups in 'of' and returns all
+  // ExprGroups that have a use in their definition of provided of IdGroups.
+  ExprGroups allUsesOf(const ValGroups& of) const;
+
+  // Recursively traverses definitions of the IdGroups in 'of' and returns all
+  // ExprGroups used in this history of defining the 'of' IdGroups.
+  ExprGroups allDefinitionsOf(const ValGroups& of) const;
+
+  //! Returns the pointer to expressions associated with the
+  //! definitions of the provided ValGroup in the provided
+  //! mapping mode (if it exists). Nullptr is returned otherwise.
+  //!
+  //! The returned pointer is to a vector of vector of expressions. The
+  //! inner vector is proven to be equivalent based on the provided mode. The
+  //! outer vector are expression groups that are not equivalent based on the
+  //! provided mode, but produce one of the ValGroups within the same disjoint
+  //! Val set based on the provided mode.
+  const ExprGroups* getDefinitions(const ValGroup& id_group) const;
+
+  //! Same as getDefinitions but for uses instead of
+  //! definitions
+  const ExprGroups* getUses(const ValGroup& id_group) const;
+
+  bool hasDefinitions(const ValGroup& id_group) const;
+
+  bool hasUses(const ValGroup& id_group) const;
+
+  std::string toString() const;
+
+  // Initializes entries for the provided Val with its definitions and
+  // uses.
+  void initializeVal(
+      Val* val,
+      const VectorOfUniqueEntries<Expr*>& definitions,
+      const VectorOfUniqueEntries<Expr*>& uses);
+
+  // Same as the above exept val->definition() and val->uses() are
+  // used
+  void initializeVal(Val* val);
+
+  // Returns true if first and second are expressions through which
+  // this ValGraph has matching inputs (if forward), or outputs (if not
+  // forward). Returning true means the expressions are "the same", in terms
+  // they modify matching original inputs by the same amount.
+  bool exprsMap(Expr* first, Expr* second, bool forward) const;
+
+  void addUniqueUses(const ValGroup& id_group, const ExprGroup& uses) {
+    unique_uses_.at(id_group).pushBack(uses);
+  }
+
+  void addUniqueDefinitions(const ValGroup& id_group, const ExprGroup& defs) {
+    unique_definitions_.at(id_group).pushBack(defs);
+  }
+
+  // Set val0 and val1 to mapped in disjointValsSet[mode], attempt to propagate
```

Comment:
```suggestion
  // Set val0 and val1 to mapped in disjointValsSet, attempt to propagate
```
Similar to my previous comment, is this a leftover of some refactor?

Reply:
Updated